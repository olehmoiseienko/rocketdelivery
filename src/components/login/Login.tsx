import React, { useState, FormEvent, ChangeEvent, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { routes } from '../../constants/routes';
import { useOrderContext } from '../../provider/OrderProvider';
import StyledLoader from '../../shared/StyledLoader';
import StyledError from '../../shared/StyledError';
import { isValidOrderNumber, isValidZipCode } from './validator';

const Login = () => {
  const navigate = useNavigate();
  const { currentOrder, isLoading, error, getOrders } = useOrderContext();
  const [orderNumber, setOrderNumber] = useState<string>('');
  const [zipCode, setZipCode] = useState<string>('');
  const [isFormValid, setIsFormValid] = useState<boolean>(false);

  const validate = () => {
    const isValidForm = isValidOrderNumber(orderNumber) && isValidZipCode(zipCode);
    setIsFormValid(isValidForm);
  };

  useEffect(() => {
    validate();
  }, [orderNumber, zipCode]);

  useEffect(() => {
    if (!error && currentOrder) {
      navigate(`${routes.orderView}`);
    }
  }, [error, currentOrder]);

  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    await getOrders({ orderNumber, zipCode });
  };

  return (
    <>
      {isLoading && <StyledLoader />}
      {error && <StyledError>{error}</StyledError>}
      <form onSubmit={(event) => handleSubmit(event)}>
        <div>
          <label htmlFor="ordernumber">Order Number</label>
          <input
            value={orderNumber}
            type="text"
            id="ordernumber"
            onChange={(e: ChangeEvent<HTMLInputElement>) => setOrderNumber(e.target.value)}
          />
        </div>

        <div>
          <label htmlFor="zipcode">Zip Code</label>
          <input
            value={zipCode}
            type="text"
            id="zipcode"
            onChange={(e: ChangeEvent<HTMLInputElement>) => setZipCode(e.target.value)}
          />
        </div>

        <button disabled={!isFormValid}>Track</button>
      </form>
    </>
  );
};

export default Login;
