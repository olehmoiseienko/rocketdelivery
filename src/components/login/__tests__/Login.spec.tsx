import React from 'react';
import { render, screen, act } from '@testing-library/react';
import { fireEvent } from '@testing-library/react';
import Login from '../../login/Login';
import { MemoryRouter } from 'react-router-dom';
import Provider from '../../../provider/OrderProvider';
import { fillInByTestId, testIds } from '../../../tests/base-helper';

jest.mock('../../../provider/OrderProvider');
const onSubmit = jest.fn();

beforeEach(() => {
  (Provider as unknown as jest.Mock).mockImplementation(() => ({
    currentOrder: null,
    getOrders: onSubmit
  }));
});

describe('Login', () => {
  it('renders', async () => {
    const tree = await render(
      <MemoryRouter initialEntries={['/']}>
        <Login />
      </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
    expect(await screen.findByText('Order Number')).toBeInTheDocument();
    expect(await screen.findByText('Zip Code')).toBeInTheDocument();
    expect(await screen.findByText('Track')).toBeInTheDocument();
    expect(await screen.findByRole('button')).toBeDisabled();
    expect(tree).toMatchSnapshot();
  });

  it('check that getOrders is called on click submit', async () => {
    await render(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    );

    await act(async () => {
      await fillInByTestId(testIds.ordernumber, '123456789');
      await fillInByTestId(testIds.zipcode, '12345');
    });

    fireEvent.click(await screen.findByRole('button'));

    expect(onSubmit).toHaveBeenCalled();
  });
});
