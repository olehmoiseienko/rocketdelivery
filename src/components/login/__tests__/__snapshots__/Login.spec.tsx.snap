// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-dkrFOg sc-hLBbgP zJZJp dpQFSz"
      >
        <span
          class="sc-eDvSVe gFwEKQ"
        >
          <span>
            <svg
              class="sc-ftTHYK cTEWnM"
              color="#333"
              data-testid="icon-Logo"
              fill="#333"
              height="100px"
              viewBox="0 0 100 100"
              width="100px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="#152A79"
                height="100"
                rx="25"
                width="100"
              />
              <path
                d="M13 73V33.5L23.5 26.5V46.5V67.5L38.5 75.5V87.5L13 73Z"
                fill="#FEFEFE"
                stroke="#FEFEFE"
              />
              <path
                d="M28 66.5L38.5 72V34H40L52.5 43V58L43 53.5V64.5L55 72L81 56V24.5L55 9L28 24.5V66.5Z"
                fill="#FEFEFE"
                stroke="#FEFEFE"
              />
              <path
                d="M69.5 34.5L57 42V57.5L70.5 49V34L69.5 34.5Z"
                fill="#152A79"
                stroke="#152A79"
              />
              <path
                d="M54 22L42 30L54 38L67 30L54 22Z"
                fill="#152A79"
                stroke="#152A79"
              />
            </svg>
          </span>
        </span>
        <h1>
          Track your order
        </h1>
        <p>
          Enter your order number and zip code combination to see the order details and shipping updates
        </p>
        <form>
          <div
            class="sc-gKPRtg hUQkau"
          >
            <label
              for="ordernumber"
            >
              Order Number
            </label>
            <input
              class="sc-jSUZER dbOMJk"
              data-testid="ordernumber"
              id="ordernumber"
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-gKPRtg hUQkau"
          >
            <label
              for="zipcode"
            >
              Zip Code
            </label>
            <input
              class="sc-jSUZER dbOMJk"
              data-testid="zipcode"
              id="zipcode"
              type="text"
              value=""
            />
          </div>
          <button
            class="sc-iBYQkv fgwwGD"
            disabled=""
          >
            <span>
              Track
            </span>
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-dkrFOg sc-hLBbgP zJZJp dpQFSz"
    >
      <span
        class="sc-eDvSVe gFwEKQ"
      >
        <span>
          <svg
            class="sc-ftTHYK cTEWnM"
            color="#333"
            data-testid="icon-Logo"
            fill="#333"
            height="100px"
            viewBox="0 0 100 100"
            width="100px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="#152A79"
              height="100"
              rx="25"
              width="100"
            />
            <path
              d="M13 73V33.5L23.5 26.5V46.5V67.5L38.5 75.5V87.5L13 73Z"
              fill="#FEFEFE"
              stroke="#FEFEFE"
            />
            <path
              d="M28 66.5L38.5 72V34H40L52.5 43V58L43 53.5V64.5L55 72L81 56V24.5L55 9L28 24.5V66.5Z"
              fill="#FEFEFE"
              stroke="#FEFEFE"
            />
            <path
              d="M69.5 34.5L57 42V57.5L70.5 49V34L69.5 34.5Z"
              fill="#152A79"
              stroke="#152A79"
            />
            <path
              d="M54 22L42 30L54 38L67 30L54 22Z"
              fill="#152A79"
              stroke="#152A79"
            />
          </svg>
        </span>
      </span>
      <h1>
        Track your order
      </h1>
      <p>
        Enter your order number and zip code combination to see the order details and shipping updates
      </p>
      <form>
        <div
          class="sc-gKPRtg hUQkau"
        >
          <label
            for="ordernumber"
          >
            Order Number
          </label>
          <input
            class="sc-jSUZER dbOMJk"
            data-testid="ordernumber"
            id="ordernumber"
            type="text"
            value=""
          />
        </div>
        <div
          class="sc-gKPRtg hUQkau"
        >
          <label
            for="zipcode"
          >
            Zip Code
          </label>
          <input
            class="sc-jSUZER dbOMJk"
            data-testid="zipcode"
            id="zipcode"
            type="text"
            value=""
          />
        </div>
        <button
          class="sc-iBYQkv fgwwGD"
          disabled=""
        >
          <span>
            Track
          </span>
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Login renders 2`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-dkrFOg sc-hLBbgP zJZJp dpQFSz"
      >
        <span
          class="sc-eDvSVe gFwEKQ"
        >
          <span>
            <svg
              class="sc-ftTHYK cTEWnM"
              color="#333"
              data-testid="icon-Logo"
              fill="#333"
              height="100px"
              viewBox="0 0 100 100"
              width="100px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="#152A79"
                height="100"
                rx="25"
                width="100"
              />
              <path
                d="M13 73V33.5L23.5 26.5V46.5V67.5L38.5 75.5V87.5L13 73Z"
                fill="#FEFEFE"
                stroke="#FEFEFE"
              />
              <path
                d="M28 66.5L38.5 72V34H40L52.5 43V58L43 53.5V64.5L55 72L81 56V24.5L55 9L28 24.5V66.5Z"
                fill="#FEFEFE"
                stroke="#FEFEFE"
              />
              <path
                d="M69.5 34.5L57 42V57.5L70.5 49V34L69.5 34.5Z"
                fill="#152A79"
                stroke="#152A79"
              />
              <path
                d="M54 22L42 30L54 38L67 30L54 22Z"
                fill="#152A79"
                stroke="#152A79"
              />
            </svg>
          </span>
        </span>
        <h1>
          Track your order
        </h1>
        <p>
          Enter your order number and zip code combination to see the order details and shipping updates
        </p>
        <form>
          <div
            class="sc-gKPRtg hUQkau"
          >
            <label
              for="ordernumber"
            >
              Order Number
            </label>
            <input
              class="sc-jSUZER dbOMJk"
              data-testid="ordernumber"
              id="ordernumber"
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-gKPRtg hUQkau"
          >
            <label
              for="zipcode"
            >
              Zip Code
            </label>
            <input
              class="sc-jSUZER dbOMJk"
              data-testid="zipcode"
              id="zipcode"
              type="text"
              value=""
            />
          </div>
          <button
            class="sc-iBYQkv fgwwGD"
            disabled=""
          >
            <span>
              Track
            </span>
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-dkrFOg sc-hLBbgP zJZJp dpQFSz"
    >
      <span
        class="sc-eDvSVe gFwEKQ"
      >
        <span>
          <svg
            class="sc-ftTHYK cTEWnM"
            color="#333"
            data-testid="icon-Logo"
            fill="#333"
            height="100px"
            viewBox="0 0 100 100"
            width="100px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              fill="#152A79"
              height="100"
              rx="25"
              width="100"
            />
            <path
              d="M13 73V33.5L23.5 26.5V46.5V67.5L38.5 75.5V87.5L13 73Z"
              fill="#FEFEFE"
              stroke="#FEFEFE"
            />
            <path
              d="M28 66.5L38.5 72V34H40L52.5 43V58L43 53.5V64.5L55 72L81 56V24.5L55 9L28 24.5V66.5Z"
              fill="#FEFEFE"
              stroke="#FEFEFE"
            />
            <path
              d="M69.5 34.5L57 42V57.5L70.5 49V34L69.5 34.5Z"
              fill="#152A79"
              stroke="#152A79"
            />
            <path
              d="M54 22L42 30L54 38L67 30L54 22Z"
              fill="#152A79"
              stroke="#152A79"
            />
          </svg>
        </span>
      </span>
      <h1>
        Track your order
      </h1>
      <p>
        Enter your order number and zip code combination to see the order details and shipping updates
      </p>
      <form>
        <div
          class="sc-gKPRtg hUQkau"
        >
          <label
            for="ordernumber"
          >
            Order Number
          </label>
          <input
            class="sc-jSUZER dbOMJk"
            data-testid="ordernumber"
            id="ordernumber"
            type="text"
            value=""
          />
        </div>
        <div
          class="sc-gKPRtg hUQkau"
        >
          <label
            for="zipcode"
          >
            Zip Code
          </label>
          <input
            class="sc-jSUZER dbOMJk"
            data-testid="zipcode"
            id="zipcode"
            type="text"
            value=""
          />
        </div>
        <button
          class="sc-iBYQkv fgwwGD"
          disabled=""
        >
          <span>
            Track
          </span>
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
